### PR TITLE
SPLAT-1293: Bump IPAM types to v1beta1.

### PR DESCRIPTION
**Pull request by vr4manta**
[SPLAT-1293](https://issues.redhat.com/browse/SPLAT-1293)

## Changes
- Updated IPAM types to v1beta1.
- Removed third-party directory data related to IPAM (installed by cluster api operator)

## Dependencies
- https://github.com/openshift/cluster-api/pull/195
- https://github.com/openshift/cluster-api/pull/197
- https://github.com/openshift/cluster-capi-operator/pull/159

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,10 +18,32 @@ ifeq ($(DBG),1)
...
+GO_INSTALL := ./scripts/go_install.sh
+
+# Helper function to get dependency version from go.mod
+get_go_version = $(shell go list -m $1 | awk '{print $$2}')
+
+#
+# Binaries.
+#
+# Note: Need to use abspath so we can invoke these from subdirectories
+CONTROLLER_GEN_VER := v0.12.0
```
Comment (JoelSpeed) [inline Makefile:38]: Are we generating the CRDs from source? Any reason not to copy from upstream files?

Do we have a check to make sure that if these are out of date, the PR gets a red status check?

**Review by JoelSpeed — COMMENTED** [review]
What is the purpose of the third_party folder if we are also importing the CAPI dependency directly? Surely we don't need the third_party folder if we are making that direct import?

We must make sure that in stable clusters we only install the v1beta1 version of the IPAM CRDs, we do not want to ship v1alpha1 and we do not want to have to run conversion webhooks.

I would also like to explore having this file come directly from the openshift Cluster API fork. I'm not sure it's required to be in this repo. The cluster API repo already ships manifests so should be able to add this one alongside

File: Makefile (hunk 1 above)
Comment (vr4manta) [reply Makefile:38]: We do have a script that syncs the CRDs.  I can update workflow to fail if CRDs are not as expected.

File: third_party/cluster-api/crd/ipam.cluster.x-k8s.io_ipaddressclaims.yaml [hunk 2]
```diff
@@ -25,6 +25,10 @@ spec:
       jsonPath: .spec.poolRef.kind
       name: Pool Kind
       type: string
+    - description: Time duration since creation of IPAdressClaim
+      jsonPath: .metadata.creationTimestamp
+      name: Age
+      type: date
     name: v1alpha1
```
[the remark this answers is not in the document]
Comment (vr4manta) [reply third_party/cluster-api/crd/ipam.cluster.x-k8s.io_ipaddressclaims.yaml:32]: v4.15 has the use of alpha.  the ipam controller for ci was using it.  Having both is allowing compatability during our ci.  the cluster-api project has both v1alpha1 and v1beta1 so I borrowed this.

**Comment by vr4manta** [comment]
> What is the purpose of the third_party folder if we are also importing the CAPI dependency directly? Surely we don't need the third_party folder if we are making that direct import?
> 
> We must make sure that in stable clusters we only install the v1beta1 version of the IPAM CRDs, we do not want to ship v1alpha1 and we do not want to have to run conversion webhooks.
> 
> I would also like to explore having this file come directly from the openshift Cluster API fork. I'm not sure it's required to be in this repo. The cluster API repo already ships manifests so should be able to add this one alongside

@JoelSpeed 

From what i see, the reason 4.15 and 4.16 broke was because cluster-api no longer install the CRDs when TechPreview is enabled.  That said, in 4.16, we want to have static IPs be GA and that will not work if cluster-api is TechPreview only and not installing CRDs. 

Currently the static IP IPAM controller used in CI needs to support 4.15 and 4.16.   4.15 was/is using v1alpha1 IPAddress/Claims.  Having these here provides that compatability without over complicating the IPAM controller for CI.

I can look into pulling the CRDs from the cluster-api fork or upstream as well.  I like the idea of generating them based on that project dependency.

File: third_party/cluster-api/crd/ipam.cluster.x-k8s.io_ipaddressclaims.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply third_party/cluster-api/crd/ipam.cluster.x-k8s.io_ipaddressclaims.yaml:32]: The IPAM controller for CI will need to be updated, we can include both for now, but if we do, do we not need to provide a conversion webhook?

To merge a PR adding both alpha and beta versions to the branches, I would want to see a release blocker bug targeted at 4.16 that prevents us shipping 4.16 without first having removed the alpha version. We do not want to ship the alpha in the stable channels

**Comment by JoelSpeed** [comment]
> From what i see, the reason 4.15 and 4.16 broke was because cluster-api no longer install the CRDs when TechPreview is enabled. That said, in 4.16, we want to have static IPs be GA and that will not work if cluster-api is TechPreview only and not installing CRDs.

@damdo is aware of the reason for this. IMO, we don't need to use the cluster-capi-operator to install CRDs, CVO can do that for us and I've asked Dam to look into this. We need to install only the IPAM related ones for now, but they can still come from the payload image for a techpreview component and operator, they just need to be tagged appropriately for the CVO to pick them up.

I think doing so would be much simpler, we can just have a script that copies the CRDs to the manifests folder and edits them as required (eg adding some annotations or dropping the alpha version). It will avoid a lot of the scripting you have here that is regenerating from source

**Comment by vr4manta** [comment]
> > From what i see, the reason 4.15 and 4.16 broke was because cluster-api no longer install the CRDs when TechPreview is enabled. That said, in 4.16, we want to have static IPs be GA and that will not work if cluster-api is TechPreview only and not installing CRDs.
> 
> @damdo is aware of the reason for this. IMO, we don't need to use the cluster-capi-operator to install CRDs, CVO can do that for us and I've asked Dam to look into this. We need to install only the IPAM related ones for now, but they can still come from the payload image for a techpreview component and operator, they just need to be tagged appropriately for the CVO to pick them up.
> 
> I think doing so would be much simpler, we can just have a script that copies the CRDs to the manifests folder and edits them as required (eg adding some annotations or dropping the alpha version). It will avoid a lot of the scripting you have here that is regenerating from source

@JoelSpeed , Excellent.  This would be super helpful for 4.15 to get this fixed.  I saw these CRD files in the CVO payload, but they are not being applied.  I've tried with CustomNoUpgrade and that conflicts with the TechPreviewNoUpgrade annotation on the file (0000_30_cluster-api_04_cm.core-cluster-api.yaml)

File: third_party/cluster-api/crd/ipam.cluster.x-k8s.io_ipaddressclaims.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (vr4manta) [reply third_party/cluster-api/crd/ipam.cluster.x-k8s.io_ipaddressclaims.yaml:32]: @JoelSpeed , I've begun updating the IPAM controller.  I have also started removing alpha from 4.16.

File: hack/crds-sync.sh
[the remark this answers is not in the document]
Comment (rvanderp3) [reply hack/crds-sync.sh:1]: we should probably remove it if it's not required

**Comment by vr4manta** [comment]
@JoelSpeed , with all third-party removed, CRDs no longer being installed via this PR, everything is looking good with the changes to the cluster-api.  https://pr-payload-tests.ci.openshift.org/runs/ci/4b5b2020-cc2a-11ee-94e9-34a677c11ec2-0

File: tools/go.mod [hunk 3]
```diff
@@ -4,47 +4,63 @@ go 1.19
 
 require (
 	github.com/ahmetb/gen-crd-api-reference-docs v0.3.0
+	github.com/mikefarah/yq/v4 v4.40.5
```
Comment (JoelSpeed) [inline tools/go.mod:7]: No longer needed?

**Review by JoelSpeed — COMMENTED** [review]
I think the YQ and tools vendor changes can also be dropped now right?

**Comment by damdo** [comment]
@vr4manta: Here is the PR to promote Core CAPI IPAM CRDs to GA: https://github.com/openshift/cluster-api/pull/197
Let me know when we can proceed merging.

File: tools/go.mod (hunk 3 above)
Comment (vr4manta) [reply tools/go.mod:7]: Good catch, I'll remove those.

**Comment by vr4manta** [comment]
@damdo , I have tested the changes with the previous PR and it looks to be installing the CRDs.  Thanks!  Do you have an image with the current removal of TechPreviewNoUpgrade so I can test real quick?  I have not built that image before and figured it might be quicker to just borrow one from you if possible.

**Comment by damdo** [comment]
@vr4manta try this `registry.build05.ci.openshift.org/ci-op-95thkpjh/release:latest`

File: tools/go.mod (hunk 3 above)
Comment (vr4manta) [reply tools/go.mod]: @JoelSpeed , Reference has been removed.

**Comment by vr4manta** [comment]
@damdo / @JoelSpeed , I ran periodical, grabbed container image, and tested locally w/ my changes with TechPreview not enabled (CI has enabled until all ducks are in the pod), and CRDs were installed and all IPAddress/Claims were initialized as expected.

**Comment by JoelSpeed** [comment]
/override security/snyk

Having reviewed the snyk scan, it's complaining about an indirect dependency that isn't actually indirect in this repository. We are not exposed since we aren't importing the package from nats at all, it's not present in our vendor.

**Pull request merged**